src/XPurchase/VirtoCommerce.ExperienceApiModule.XOrder/Commands/AddOrUpdateOrderPaymentCommandHandler.cs

```diff
@@ -0,0 +1,64 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using VirtoCommerce.OrdersModule.Core.Services;
+using VirtoCommerce.PaymentModule.Core.Model;
+using VirtoCommerce.PaymentModule.Core.Model.Search;
+using VirtoCommerce.PaymentModule.Core.Services;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.GenericCrud;
+
+namespace VirtoCommerce.ExperienceApiModule.XOrder.Commands
+{
+    public class AddOrUpdateOrderPaymentCommandHandler : IRequestHandler<AddOrUpdateOrderPaymentCommand, CustomerOrderAggregate>
+    {
+        private readonly ICustomerOrderAggregateRepository _customerOrderAggregateRepository;
+        private readonly ICustomerOrderService _customerOrderService;
+        private readonly ISearchService<PaymentMethodsSearchCriteria, PaymentMethodsSearchResult, PaymentMethod> _paymentMethodsSearchService;
+
+        public AddOrUpdateOrderPaymentCommandHandler(
+            ICustomerOrderAggregateRepository customerOrderAggregateRepository,
+            ICustomerOrderService customerOrderService,
+            IPaymentMethodsSearchService paymentMethodsSearchService)
+        {
+            _customerOrderAggregateRepository = customerOrderAggregateRepository;
+            _customerOrderService = customerOrderService;
+            _paymentMethodsSearchService = (ISearchService<PaymentMethodsSearchCriteria, PaymentMethodsSearchResult, PaymentMethod>)paymentMethodsSearchService;
+        }
+
+        public async Task<CustomerOrderAggregate> Handle(AddOrUpdateOrderPaymentCommand request, CancellationToken cancellationToken)
+        {
+            var orderAggregate = await _customerOrderAggregateRepository.GetOrderByIdAsync(request.OrderId);
+
+            var paymentId = request.Payment.Id?.Value ?? null;
+            var payment = orderAggregate.Order.InPayments.FirstOrDefault(s => paymentId != null && s.Id == paymentId);
+            payment = request.Payment.MapTo(payment);
```

Comment:
Should we check payment for null before `MapTo(payment)` call?

Reply:
payment  will be created in MapTo if null